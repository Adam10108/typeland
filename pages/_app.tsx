import Head from 'next/head'
import '../styles/base.css'

export default function MyApp({
  Component,
  pageProps,
}: {
  Component: any
  pageProps: any
}): JSX.Element {
  return (
    <>
      <Head>
        <title>Typeland</title>
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <Component {...pageProps} />
    </>
  )
}
