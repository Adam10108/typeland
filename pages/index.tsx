/** @jsx jsx */
import { css, jsx } from '@emotion/core'
import tw from '@tailwindcssinjs/macro'
import { useState } from 'react'

const s = {
  container: tw`
    h-screen min-w-full
    flex
    items-center justify-center
  `,
  title: tw`
    text-4xl
    text-blue-600
  `,
  typingArea: tw`
    text-lg
  `,
  typingInput: tw`
    h-0
    w-0
  `,
  correct: tw`
    text-green-600
  `,
  wrong: tw`
    text-red-400
  `,
  caret: [
    tw`
    h-4
      border border-blue-500
    `,
    css`
      width: 1px;
    `,
  ],
}

const TypingArea = ({ text, userText, showCaret }): JSX.Element => {
  return (
    <p css={s.typingArea}>
      {text}
      <br />
      {userText.map((text, i) => {
        return i % 2 == 0 ? (
          <span
            css={s.correct}
            key={i}
            className="correct"
            data-testid="correct"
          >
            {text}
          </span>
        ) : (
          <span css={s.wrong} key={i} className="wrong" data-testid="wrong">
            {text}
          </span>
        )
      })}
      {showCaret ? <span css={s.caret} data-testid="caret"></span> : null}
    </p>
  )
}

export const Home = (): JSX.Element => {
  const textToType = 'the quick brown fox jumps over the lazy dog'
  const [userTypeInput, setUserTypeInput] = useState([''])
  const [inputIsFocused, setInputIsFocused] = useState(true)

  const handleDelete = () => {
    let newUserTypeInput = [...userTypeInput]

    newUserTypeInput[newUserTypeInput.length - 1] = newUserTypeInput[
      newUserTypeInput.length - 1
    ].slice(0, -1)

    if (
      newUserTypeInput[newUserTypeInput.length - 1].length == 0 &&
      userTypeInput.length > 1
    ) {
      newUserTypeInput = newUserTypeInput.slice(0, -1)
    }

    setUserTypeInput(newUserTypeInput)
  }

  const handleType = (e) => {
    const { key, keyCode } = e

    if (keyCode == 8) {
      return handleDelete()
    }

    // Filter out modifiers
    if (key.length != 1) {
      return
    }

    const userTypeLength = userTypeInput.reduce((p, c) => p + c.length, 0)

    let newUserTypeInput = [...userTypeInput]

    if (textToType[userTypeLength] == key && newUserTypeInput.length % 2 == 0) {
      newUserTypeInput = newUserTypeInput.concat('')
    } else if (
      textToType[userTypeLength] != key &&
      newUserTypeInput.length % 2 == 1
    ) {
      newUserTypeInput = newUserTypeInput.concat('')
    }

    newUserTypeInput[newUserTypeInput.length - 1] = `${
      newUserTypeInput[newUserTypeInput.length - 1]
    }${key}`

    setUserTypeInput(newUserTypeInput)
  }

  return (
    <div
      css={s.container}
      onClick={() => document.getElementById('typingInput').focus()}
    >
      <main>
        <h1 css={s.title}>Typeland</h1>

        <TypingArea
          text={textToType}
          userText={userTypeInput}
          showCaret={inputIsFocused}
        />

        <input
          id="typingInput"
          data-testid="typingInput"
          css={s.typingInput}
          type="text"
          ref={(i) => i && i.focus()} // Autofocus
          onKeyDown={handleType}
          onBlur={() => {
            setInputIsFocused(false)
          }}
          onFocus={() => {
            setInputIsFocused(true)
          }}
          tabIndex={0}
        />
      </main>
    </div>
  )
}

export default Home
