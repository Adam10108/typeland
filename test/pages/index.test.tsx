import React from 'react'
import { render, fireEvent } from '../testUtils'
import { Home } from '../../pages/index'

describe('Home page', () => {
  it('renders title', () => {
    const { asFragment, getByText } = render(<Home />, {})
    getByText(/Typeland/i)

    expect(asFragment()).toMatchSnapshot()
  })

  it('renders text to type', () => {
    const { getByText } = render(<Home />, {})

    getByText(/the quick brown fox jumps over the lazy dog/i)
  })

  it('highlights text which is typed', () => {
    const { getByTestId } = render(<Home />, {})

    const typingInput = getByTestId('typingInput')

    fireEvent.keyDown(typingInput, { key: 't' })
    expect(getByTestId('correct')).toHaveTextContent('t')

    fireEvent.keyDown(typingInput, { key: 'h' })
    expect(getByTestId('correct')).toHaveTextContent('th')

    fireEvent.keyDown(typingInput, { key: 'x' })
    expect(getByTestId('correct')).toHaveTextContent('th')
    expect(getByTestId('wrong')).toHaveTextContent('x')
  })
})
