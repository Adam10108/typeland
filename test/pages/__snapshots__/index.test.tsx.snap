// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page can delete typed text with backspace 1`] = `
<DocumentFragment>
  .emotion-5 {
  height: 100vh;
  min-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0 {
  --text-opacity: 1;
  font-size: 2.25rem;
  color: #1c64f2;
  color: rgba(28,100,242,var(--text-opacity));
}

.emotion-3 {
  font-size: 1.125rem;
}

.emotion-1 {
  --text-opacity: 1;
  color: #057a55;
  color: rgba(5,122,85,var(--text-opacity));
}

.emotion-2 {
  --border-opacity: 1;
  height: 1rem;
  border-width: 1px;
  border-color: #3f83f8;
  border-color: rgba(63,131,248,var(--border-opacity));
  width: 1px;
}

.emotion-4 {
  height: 0;
  width: 0;
}

<div
    class="emotion-5"
  >
    <main>
      <h1
        class="emotion-0"
      >
        Typeland
      </h1>
      <p
        class="emotion-3"
      >
        the quick brown fox jumps over the lazy dog
        <br />
        <span
          class="correct emotion-1"
          data-testid="correct"
        >
          th
        </span>
        <span
          class="emotion-2"
          data-testid="caret"
        />
      </p>
      <input
        class="emotion-4"
        data-testid="typingInput"
        id="typingInput"
        tabindex="0"
        type="text"
      />
    </main>
  </div>
</DocumentFragment>
`;

exports[`Home page highlights text which is typed 1`] = `
<DocumentFragment>
  .emotion-6 {
  height: 100vh;
  min-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0 {
  --text-opacity: 1;
  font-size: 2.25rem;
  color: #1c64f2;
  color: rgba(28,100,242,var(--text-opacity));
}

.emotion-4 {
  font-size: 1.125rem;
}

.emotion-1 {
  --text-opacity: 1;
  color: #057a55;
  color: rgba(5,122,85,var(--text-opacity));
}

.emotion-3 {
  --border-opacity: 1;
  height: 1rem;
  border-width: 1px;
  border-color: #3f83f8;
  border-color: rgba(63,131,248,var(--border-opacity));
  width: 1px;
}

.emotion-5 {
  height: 0;
  width: 0;
}

.emotion-2 {
  --text-opacity: 1;
  color: #f98080;
  color: rgba(249,128,128,var(--text-opacity));
}

<div
    class="emotion-6"
  >
    <main>
      <h1
        class="emotion-0"
      >
        Typeland
      </h1>
      <p
        class="emotion-4"
      >
        the quick brown fox jumps over the lazy dog
        <br />
        <span
          class="correct emotion-1"
          data-testid="correct"
        >
          th
        </span>
        <span
          class="wrong emotion-2"
          data-testid="wrong"
        >
          x
        </span>
        <span
          class="emotion-3"
          data-testid="caret"
        />
      </p>
      <input
        class="emotion-5"
        data-testid="typingInput"
        id="typingInput"
        tabindex="0"
        type="text"
      />
    </main>
  </div>
</DocumentFragment>
`;

exports[`Home page renders title 1`] = `
<DocumentFragment>
  .emotion-5 {
  height: 100vh;
  min-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0 {
  --text-opacity: 1;
  font-size: 2.25rem;
  color: #1c64f2;
  color: rgba(28,100,242,var(--text-opacity));
}

.emotion-3 {
  font-size: 1.125rem;
}

.emotion-1 {
  --text-opacity: 1;
  color: #057a55;
  color: rgba(5,122,85,var(--text-opacity));
}

.emotion-2 {
  --border-opacity: 1;
  height: 1rem;
  border-width: 1px;
  border-color: #3f83f8;
  border-color: rgba(63,131,248,var(--border-opacity));
  width: 1px;
}

.emotion-4 {
  height: 0;
  width: 0;
}

<div
    class="emotion-5"
  >
    <main>
      <h1
        class="emotion-0"
      >
        Typeland
      </h1>
      <p
        class="emotion-3"
      >
        the quick brown fox jumps over the lazy dog
        <br />
        <span
          class="correct emotion-1"
          data-testid="correct"
        />
        <span
          class="emotion-2"
          data-testid="caret"
        />
      </p>
      <input
        class="emotion-4"
        data-testid="typingInput"
        id="typingInput"
        tabindex="0"
        type="text"
      />
    </main>
  </div>
</DocumentFragment>
`;
